k9mail/src/main/java/com/fsck/k9/fragment/MessageListFragment.java

```diff
@@ -2719,6 +2733,113 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
 
             fragmentListener.updateMenu();
         }
+
+        categorizedMessages = getMessageCategories();
+    }
+
+    int getDateCategory(long date) {
+        /*
+         * Outlook descriptions "Today", "Yesterday", the previous days of the
+         * week ending with Sunday, "Last Week", "Two Weeks Ago",
+         * "Three Weeks Ago", "Earlier this Month", "Last Month", "Older"
+         */
+
+        if (date != 0) {
+            int category = 0;
+            Calendar calendar = Calendar.getInstance();
+
+            int sundayOffset = 0;
+            int thisMonth = calendar.get(calendar.MONTH);
+            for (int j = 0; j < 7; j++) {
+                calendar.add(Calendar.DATE, -1);
+                if (calendar.get(calendar.DAY_OF_WEEK) == calendar.SUNDAY) {
+                    sundayOffset = j + 1;
+                    break;
+                }
+            }
+
+            Calendar messageDate = Calendar.getInstance();
+            messageDate.setTimeInMillis(date);
+
+            // calendar.setTime( compareDate );
+            long difference = daysBetween(messageDate, Calendar.getInstance());
+
+            // dateDifference( millisCompare, millisNow );
+            if (difference == 0) {
+                category = 1;
+            } else if (difference == 1) {
+                category = 2;
+            } else if (difference == 2) {
+                category = 3;
+            } else if ((sundayOffset > 1) && (difference <= sundayOffset)) {
+                category = messageDate.get(Calendar.DAY_OF_WEEK) + 3;
+            } else if ((difference - sundayOffset) < 7) {
+                category = 11;
+            } else if ((difference - sundayOffset) < 14) {
+                category = 12;
+            } else if ((difference - sundayOffset) < 21) {
+                category = 13;
+            } else {
+                category = 14;
+            }
+
+            return category;
+        }
+        return 0;
+    }
+
+    private final static long MILLISECS_PER_DAY = 24 * 60 * 60 * 1000;
+
+    private long daysBetween(Calendar a, Calendar b) {
```

Comment:
Belongs in a helper class (or ideally, just use a date library like `joda-time-android`)